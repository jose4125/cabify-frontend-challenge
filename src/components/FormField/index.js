import React from 'react';

import Field from '../Field';
import warning from '../../images/warning.svg';

const FormField = ({
  name,
  label,
  col,
  isValid,
  touched,
  isDirty,
  isFocus,
  ...props
}) => {
  let active = (isFocus && touched) || (touched && isDirty);
  const focus = isFocus;
  const error =
    (isValid !== undefined &&
      isValid === false &&
      touched &&
      isFocus === false) ||
    (isValid !== undefined && isValid === false && touched && isDirty);
  return (
    <div
      className={`formField-input ${active ? 'active ' : ''}${
        focus ? 'focus ' : ''
      }${error ? 'error ' : ''}col ${col ? `col${col}` : 'col12'}`}
    >
      <div className="input">
        <Field {...props} />
        {error && <img className="warning" src={warning} alt="warning" />}
        <label htmlFor={name}>{label}</label>
      </div>
    </div>
  );
};

export default FormField;
