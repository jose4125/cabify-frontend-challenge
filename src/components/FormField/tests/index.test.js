import React from 'react';
import { shallow } from 'enzyme';

import FormField from '../index';

describe('FormField', () => {
  const props = {
    name: 'fullname',
    stateName: 'fullName',
    label: 'Full name',
    name: 'fullname',
    label: 'Full name',
    value: '',
    isValid: false,
    touched: false,
    isDirty: false,
    isFocus: false
  };

  const wrapper = shallow(<FormField {...props} />);
  it('Should render with out crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should have a Field component', () => {
    expect(wrapper.find('Field').exists()).toBe(true);
  });

  it('should be on default mode', () => {
    expect(wrapper.find('.formField-input').prop('className')).toContain(
      'formField-input col col12'
    );
  });

  it('should not have warning icon on default mode', () => {
    expect(wrapper.find('.warning').exists()).toBe(false);
  });

  it('should have a label', () => {
    expect(wrapper.find('label').text()).toBe(props.label);
  });

  describe('FormField classes', () => {
    it('should have active and focus classes when the field is foucs', () => {
      const props = {
        isValid: false,
        touched: true,
        isDirty: false,
        isFocus: true
      };

      const wrapper = shallow(<FormField {...props} />);
      expect(wrapper.find('.formField-input').prop('className')).toContain(
        'active focus'
      );
    });

    describe('FormField valid value', () => {
      const props = {
        isValid: true,
        touched: true,
        isDirty: true,
        isFocus: false
      };

      const wrapper = shallow(<FormField {...props} />);

      it('should have active class when the field has a value and was blurred', () => {
        expect(wrapper.find('.formField-input').prop('className')).toContain(
          'active'
        );
      });

      it('should not have warning icon when the field has a correct value', () => {
        expect(wrapper.find('.warning').exists()).toBe(false);
      });
    });

    describe('FormField invalid value', () => {
      const props = {
        isValid: false,
        touched: true,
        isDirty: true,
        isFocus: false
      };

      const wrapper = shallow(<FormField {...props} />);

      it('should have error class when the field has an incorrect value', () => {
        expect(wrapper.find('.formField-input').prop('className')).toContain(
          'active error'
        );
      });

      it('should have warning icon when the field has an incorrect value', () => {
        expect(wrapper.find('.warning').exists()).toBe(true);
      });
    });

    describe('FormField required', () => {
      const props = {
        isValid: false,
        touched: true,
        isDirty: false,
        isFocus: false
      };

      const wrapper = shallow(<FormField {...props} />);

      it('should have error class when the field has not a value and was blurred', () => {
        expect(wrapper.find('.formField-input').prop('className')).toContain(
          'error'
        );
      });

      it('should have warning icon when the field is required and it doesn`t have a value', () => {
        expect(wrapper.find('.warning').exists()).toBe(true);
      });
    });

    describe('FormField col', () => {
      const props = { col: '9' };
      const wrapper = shallow(<FormField {...props} />);
      it('should have a specific col number class', () => {
        expect(wrapper.find('.formField-input').prop('className')).toContain(
          `col${props.col}`
        );
      });
    });
  });
});
