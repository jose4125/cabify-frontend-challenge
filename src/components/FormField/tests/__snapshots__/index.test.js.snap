// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField Should render with out crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FormField
    isDirty={false}
    isFocus={false}
    isValid={false}
    label="Full name"
    name="fullname"
    stateName="fullName"
    touched={false}
    value=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="input"
      >
        <Field
          stateName="fullName"
          value=""
        />
        <label
          htmlFor="fullname"
        >
          Full name
        </label>
      </div>,
      "className": "formField-input col col12",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Field
            stateName="fullName"
            value=""
          />,
          false,
          <label
            htmlFor="fullname"
          >
            Full name
          </label>,
        ],
        "className": "input",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "stateName": "fullName",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Full name",
            "htmlFor": "fullname",
          },
          "ref": null,
          "rendered": "Full name",
          "type": "label",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="input"
        >
          <Field
            stateName="fullName"
            value=""
          />
          <label
            htmlFor="fullname"
          >
            Full name
          </label>
        </div>,
        "className": "formField-input col col12",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Field
              stateName="fullName"
              value=""
            />,
            false,
            <label
              htmlFor="fullname"
            >
              Full name
            </label>,
          ],
          "className": "input",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "stateName": "fullName",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Full name",
              "htmlFor": "fullname",
            },
            "ref": null,
            "rendered": "Full name",
            "type": "label",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": false,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
